### PR TITLE
Add possibility to use large strings in WebPage::runJavaScript without copies

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### 2830ec003bed1b2bd596d3e2c5d337ee49813112
<pre>
Add possibility to use large strings in WebPage::runJavaScript without copies
<a href="https://bugs.webkit.org/show_bug.cgi?id=303295">https://bugs.webkit.org/show_bug.cgi?id=303295</a>
<a href="https://rdar.apple.com/165603388">rdar://165603388</a>

Reviewed by Ben Nham.

Add IPC::TransferString to support explicit serialization of large
strings. In subsequent patches, this will be used to pass large JS
strings as memory that is shared between multiple WebContent processes.

Avoids extra NSString * -&gt; wtf::String copy that was done only to
send the string via IPC.

To be conservative, the feature is added as explicit TransferString
instead of being built-in to IPC string serialization. If the feature
turns out to be non-problematic and receive-side API issues are solved,
it may be merged as the default String serialization.

Reland:
Original reverted in 304433@main.
Relanding with null-checking added for cases where client calls
-[WKWebView _evaluateJavaScript:...] with
NSString *javaScriptString == nil.

Add a test that evaluating @&quot;&quot; and nil behaves the same, similar to
before this patch.

* Source/JavaScriptCore/API/JSStringRefCF.cpp:
(JSStringCreateWithCFString):
* Source/WTF/wtf/cf/VectorCF.h:
(WTF::CFStringGetLatin1CStringSpan):
(WTF::CFStringGetCharactersSpan):
(WTF::CFStringCopyCharactersSpan):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::sizeInBytes const):
* Source/WTF/wtf/text/cf/AtomStringImplCF.cpp:
(WTF::AtomStringImpl::add):
* Source/WTF/wtf/text/cf/StringCF.cpp:
(WTF::String::String):
* Source/WTF/wtf/text/cf/StringConcatenateCF.h:
(WTF::StringTypeAdapter&lt;CFStringRef&gt;::writeTo&lt;char16_t&gt; const):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/Sources.txt:
* Source/WebKit/Scripts/webkit/opaque_ipc_types.tracking.in:
* Source/WebKit/Shared/RunJavaScriptParameters.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageEvaluateJavaScriptInFrame):
(callAsyncJavaScript):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(unknownError):
(-[WKWebView _evaluateJavaScript:asAsyncFunction:withSourceURL:withArguments:forceUserGesture:inFrame:inWorld:completionHandler:]):
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::runScript):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):

Canonical link: <a href="https://commits.webkit.org/304521@main">https://commits.webkit.org/304521@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/249a00937bf4e560d71b43a9f6e3efdab822b5dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135853 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8229 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47145 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143566 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [loading 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ec7dc339-663d-41d1-87c5-29946fbef236) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137722 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8075 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103823 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1d4a747f-7182-42da-9ea0-b4e0c0161b4a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138799 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/6429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121768 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84700 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/196f044d-204c-4707-bbe7-bee6ea2f1188) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/135201 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3776 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4170 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127832 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39958 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146317 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134337 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7914 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40523 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112180 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7940 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/6644 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112566 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28559 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/6047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118068 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61820 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7961 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167144 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7694 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71512 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43629 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7915 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7777 "Built successfully") | | | |

### CODE REVIEW

File: Source/WTF/wtf/cf/VectorCF.h [hunk 1]
```diff
@@ -156,12 +156,27 @@ inline std::span<const char> CFStringGetASCIICStringSpan(CFStringRef string)
...
+    auto* characters = CFStringGetCharactersPtr(string);
+    if (!characters)
+        return { };
+    return unsafeMakeSpan(reinterpret_cast<const char16_t*>(characters), CFStringGetLength(string));
+}
+
+inline void CFStringCopyCharactersSpan(CFStringRef string, std::span<char16_t> span)
+{
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    CFStringGetCharacters(string, CFRangeMake(0, span.size()), reinterpret_cast<UniChar*>(span.data()));
```
Comment (achristensen07) [inline Source/WTF/wtf/cf/VectorCF.h]: Do we need to check that the length of the span is less than or equal to the length of the string?

File: Source/WebKit/Platform/IPC/TransferString.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        },
+        [&](std::span<const char16_t> characters) {
+            m_storage = String { characters };
+        },
+        [&](SharedSpan8 handle) {
+            m_storage = WTFMove(handle);
+        },
+        [&](SharedSpan16 handle) {
+            m_storage = WTFMove(handle);
+        });
```
Comment (achristensen07) [inline Source/WebKit/Platform/IPC/TransferString.h]: nit: I think it would look better to have the ) on the next line.

**Review by achristensen07 — COMMENTED** [review]
This uses TransferStrings for all JS evaluation, and it doesn't cache them.  I'm not sure that's a direction we want to head in.

**Comment by bnham** [comment]
> This uses TransferStrings for all JS evaluation, and it doesn't cache them. I'm not sure that's a direction we want to head in.

I don't think there's an actual change in behavior expected from just this patch. If you look at `TransferString::create(CFStringRef)`, we're still creating a copy of the string when we send it across the wire. This is equivalent to what we were doing before, except that we were sending over a String instead of a TransferString.

File: Source/WTF/wtf/cf/VectorCF.h (hunk 1 above)
Comment (bnham) [reply Source/WTF/wtf/cf/VectorCF.h]: Hm yeah, I thought CF asserted if the range didn't make sense, but it looks like it doesn't.

**Review by bnham — APPROVED** [review]
From what I can tell, this will cause both empty and nil strings on the sender side to still be deserialized as nil `WTF::String` instances on the receiver side at `TransferString::release()` time.

This seems to work okay for current use case, but I wonder if it makes sense to add an empty string variant so that sending an empty string results in an empty string on the receiver side rather than a nil string?
